Typeahead fixes and enhancements

johnyanarella opened the pull request:
#### General fixes and enhancements:
- Added pass-through of `alwaysFloatLabel` property to input container.
- Tweaked CSS to ensure the input text doesn't overflow the dropdown arrow.
- Added optional clear button.
- Tweaked logic for scrolling the dropdown during keyboard navigation.
#### Typeahead (and combobox) mode fixes and enhancements:
- Disabled browser autocomplete (otherwise the native browser autocomplete overlaps this component's dropdown).
- Fixed a bug where clicking on a dropdown option was ignored if the input had focus.
  - The `blur` handler was firing (clearing the result) before the `tap` handler.
  - Changed this to use a `down` handler, since that has higher priority.
  - Replaced component-level `tap` handling via `updateSelected()` with item level `tap` handling via `_selectItem()`.
- Consolidated common logic in `setValue()`, `setValueFromLabel()` and `clearSelection()` to avoid code duplication and ensure consistent UI state changes for all combo types.
- Added `_findFilteredOptionLabelIndex` and `_findFilteredOptionValueIndex` to allow the correct match to be highlighted (and items to be unhighlighted) while typing in the typeahead.
- Changed selector selected state to be lazily initialized/updated when the dropdown is opened and when the input is changed.
- Added "No matches" item (with configurable text) for typeaheads.
- Tweaked keyboard navigation - only commits value on `Enter` or when dropdown is closed.
  - Previously silently updated `value` during interaction.
  - Now updates label and applies to `value` (and notifies via event) when interaction is complete.
  - This helped resolve typeahead issues with sending duplicate notifications (or failing to send them).

![radium-combo-enhancements-optimized](https://cloud.githubusercontent.com/assets/140385/14189865/9ac40c8a-f75d-11e5-84fb-4de7890e0df2.gif)

Commit:
Fix to disable browser autocomplete.

Commit:
Fix to ensure input text doesn't overflow over dropdown arrow.

Commit:
Typeahead mode dropdown interaction fixes.

* Fixed a bug where clicking on a typeahead option was ignored if the input had focus.
  * The blur handler fires (clearing the result) before the tap handler (which would set it to the new value) - changed this to use a down handler, since that has higher priority.
  * Replaced component level tap `_updateSelected` observer with item level `_selectItem` handler.
* Consolidated common logic in `setValue`, `setValueFromLabel` and `clearSelection` to ensure the UI is set to the correct state for all combo types.
* Adde `_findFilteredOptionLabelIndex` and `_findFilteredOptionValueIndex` to allow the correct item to be highlighted (and items to be unhighlighted) while typing in the typeahead.
* Changed selector selected state to be lazily initialized when the dropdown is opened, and when the input is changed.

Commit:
Added optional clear button and "No matches" typeahead message.

* Added "No matches" item (with configurable text) for typeaheads.
* Added optional clear button for all combo types.
* Added "Show Clear Button" toggle to demo page.
* Added `paper-toggle-button` to bower dev dependencies.

Commit:
Added pass-through of alwaysFloatLabel property to input container.

Commit:
Fix selection indicator bug introduced by "No Matches" option.

Polymer dom-if template was just hiding it, rather than removing it from the DOM. This affected the item indices (off-by-one).

Commit:
Fix for keyboard navigation - only commit value on Enter or close.

No longer silently changes the underlying value as the user keyboard navigates up / down through the dropdown - updates the label and later applies the value (and notifies via event) when the interaction is completed via Enter key or dropdown is closed.

Commit:
Fixes for dropdown scrolling during keyboard navigation.

Commit:
Fix for extra notification when in typeahead mode.

Commit:
Tweaked previous fix for combobox mode.

jasongardnerlv commented:
I'm seeing a few issues as I play with the demos.
1. The dropdown is covering up the input box for me.
   <img width="676" alt="screen shot 2016-04-03 at 2 38 38 pm" src="https://cloud.githubusercontent.com/assets/1305188/14235124/e6704a7e-f9a9-11e5-9d1b-ea257ef96df5.png">
2. The combobox is no longer sending events when applying an input value that is not in the list, unless it is immediately following a list selection.  Example:  click the combobox and select "Steve".  The event fires. Then, remove the last "e" and hit enter.  An event goes out for "Stev".  Remove the "v" and see that no event is fired.
3. On the typeahead, select "Mary".  Tab to the combobox.  Now, shift-tab back to the typeahead.  Press the right arrow to move to the end.  Backspace over the "y".  The list shows all items when it should only show "Mary", since the text in the input is "Mar".

Commit:
Fix for failure to send events on subsequent non-list combo entries.

The culprit was a bad equality test for the new value vs old value in `_updateAndDispatch()`.

When multiple non-list combo box entries are set in sequence, the new `__value` will be the same as the old `__value` (a value of `undefined`) even though the `__label` changed. The old equality test treated them as equal and incorrectly skipped the update and notification logic.

Commit:
Fix to ensure re-filtering of the list when input doesn't match an item.

When opening the dropdown, filters the options if the input doesn't match the currently selected value's label (and its not a non-list value).

Think this maintaines the delicate balance here to retain the behavior of showing the whole list if an item has been selected (vs a partial value being typed and the list being filtered).

Commit:
Tweak to previous commit - fix for typeahead mode.

The typeahead beahavior which was broken by the previous commit.

johnyanarella commented:
Good catch! I just pushed fixes for the second and third issues.

The first issue is related to `iron-dropdown` which changed how it positions menus since the `1.0.6` release we're both using in our shared internal projects. (The `bower.json` in this project references `^1.0.6`, so it pulls in a newer version.)

I'll investigate if there's a positioning tweak I can find that works for both versions of `iron-dropdown`.

johnyanarella commented:
Gross - you can see `paper-dropdown-menu` currently exhibits the same overlapping menu behavior when using the newer `iron-dropdown` version.

https://elements.polymer-project.org/elements/paper-dropdown-menu?view=demo:demo/index.html&active=paper-dropdown-menu

johnyanarella commented:
So, when using newer releases of `iron-dropdown`, we can position the menu below the input by adding the following to `_openDropdown()`:

``` javascript
this.$.dropdown.verticalOffset = 60;
```

However, with the `1.0.6` release referenced by your ambrosia library, this ends up pushing the dropdown an extra 60 pixels too far down.

Pondering a DOM measurement check to detect whether to apply a vertical offset.

johnyanarella commented:
DOM measurement as a way of detecting whether the menu is incorrectly positioned does not seem viable because the menu is not positioned during the `open()` call but asynchronously at some unknown future turn of the event loop. (I tried `this.async(..., 1)` and measurements were not available, `this.async(..., 100)` and measurements were available.)

Commit:
Fix for menu overlap - conditional workaround when problem detected.

johnyanarella commented:
I turns out that when the `iron-overlay-opened` event is fired, the coordinates have been set. Added a handler for that event that performs DOM measurement to determine if the menu position needs to be adjusted.

Works for the version of `iron-dropdown` we're using in ambrosia and with the latest version.

jasongardnerlv commented:
Approved!
